Get gdsfactory PCells working on KLayout

SkandanC opened the pull request:
We can now use gdsfactory generic pdk (or any other pdk derived from it), straight in KLayout.

Major stuff:
- Combined `link_gdsfactory_python` and `import_generic_pcells`.
- Added a class to turn gdsfactory Components to Klayout PCells
- Progress bar to show component loading progress
- Fix bug preventing update of PCells without changing name
        - This was an issue on gdsfactory's side, doing gf.clear_cache() fixed it

tvt173 commented:
Nice work @SkandanC ! Out of curiosity, what is required to enable this functionality for a PDK _not_ derived from the generic PDK?

And how do the unsupported arguments get treated now? For example, are they greyed out and somehow shown to exist but not be editable? Or are they dropped from the display entirely?

SkandanC commented:
> Nice work @SkandanC ! Out of curiosity, what is required to enable this functionality for a PDK _not_ derived from the generic PDK?
> 
> And how do the unsupported arguments get treated now? For example, are they greyed out and somehow shown to exist but not be editable? Or are they dropped from the display entirely?

Hi @tvt173,

As long as your PDK's components are gdsfactory Components, this should work as long as you modify the macro such that the currect active pdk is your pdk of interest. The class in the macro is "universal" aka it should work for any gdsfactory Component, and turn it into a KLayout PCell.

The unsupported arguments are just omitted entirely. I would like to work on this when I get time to add support for these too, that way we can instantiate, for example, an array of mzis through the `array` PCell (rather than the default way now, where it uses unmodifiable Straights).

joamatab commented:
great work Skandan! thank you

@lukasc-ubc

SkandanC commented:
It should work on Linux, and MacOS. Currently there is an issue with getting one of our dependencies, gdstk, to be imported in KLayout on Windows. [But we are working on it.](https://github.com/gdsfactory/gdsfactory/issues/891)

thomasdorch reviewed (commented):
Why was the target directory changed from `salt` to `macros`? The point of the function is to install the whole `gdsfactory/klayout` directory into KLayout's `salt` package manager so that macros, technology, and cross section scripts are all bundled under the same directory.

SkandanC commented:
Hi @thomasdorch,

I just changed the script to install to `salt` directory, and I could not see it in the tool bar for some reason.
![image](https://user-images.githubusercontent.com/100642027/203125614-f956c81e-b97b-41d5-a102-78d0d4f51515.png)

I changed it back to `macros` and I see this:
![image](https://user-images.githubusercontent.com/100642027/203125372-a3819c9e-95ff-4ba3-9262-cda374646577.png)

Maybe I am doign something wrong here, if so please let me know.

thomasdorch commented:
Is it making a symlink or a junction?

SkandanC commented:
It does not show under "Current Packages". This is on WSL, if that makes any difference. Perhaps the script is creating a junction, even though it says symlink?
